File: services/core/internal/delivery/completion_test.go
package delivery

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strings"
	"testing"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/DulsaraNethmin/Shipper/services/core/internal/clock"
	"github.com/DulsaraNethmin/Shipper/services/core/internal/httpx"
	"github.com/DulsaraNethmin/Shipper/services/core/internal/testsupport/pgtest"
)

// SHIP-123 — the two facts a delivered job carries that nothing held until now.
//
// # What this closes, and it had been open since SHIP-118 named it
//
// `Docs/01` §4.4 lists four things a delivered job requires: recipient name, delivery timestamp,
// delivery note, and photo proof. Three of the four were reachable — the timestamp is
// `actor_recorded_at`, and SHIP-115/116/118 made the evidence recordable and mandatory. **No column
// held the recipient or the note**, `000605`'s own comment named SHIP-123 as the ticket that would
// close it, and `Docs/11` §4 has carried the gap as documentation ahead of code ever since.
//
// # The rule is in both directions, and the second half is the one worth testing
//
// Required on 'Delivered' is what the document asks for. **Refused everywhere else** is what stops
// the columns becoming a general-purpose pair somebody attaches to a pickup, which would be
// recording a handover that did not happen. Both halves live in the domain — where a client is told
// which field is wrong — and in `ck_milestones_delivery_details`, where the rule stops depending on
// which function did the writing. Every test below drives one of the two layers deliberately.

// deliverable is a job driven to 'In transit' by its provider, ready for one delivery.
//
// Built through the service rather than in SQL, so that the milestone rows before the delivery are
// real ones — a fixture that moved the job with an UPDATE would leave the transition guard
// untested and, since SHIP-123, would also miss that the three earlier milestones are the ones the
// constraint has to *refuse* the new fields on.
func deliverable(t *testing.T, pool *pgxpool.Pool, svc *Service, provider, jobID uuid.UUID) {
	t.Helper()

	for i, milestone := range []Milestone{MilestoneEnRouteToPickup, MilestonePickedUp, MilestoneInTransit} {
		if _, _, err := recordMilestone(t, pool, svc, provider, jobID,
			Recording{Milestone: milestone, Key: "deliverable-" + string(rune('a'+i))}); err != nil {
			t.Fatalf("driving the job to %s: %v", milestone, err)
		}
	}
}

// TestADeliveredMilestoneCarriesTheRecipientAndTheNote is the *Done when*'s first clause.
//
// The values are read back out of the row rather than off the response, because the response is
// rendered from what the service returned and the row is what a customer's tracking view, a
// dispute and an administrator will actually read.
func TestADeliveredMilestoneCarriesTheRecipientAndTheNote(t *testing.T) {
	pool := pgtest.DB(t)
	customer := newAccount(t, pool, "comp1-c@example.com", "+61400000770", "customer")
	provider := newAccount(t, pool, "comp1-p@example.com", "+61400000771", "provider")
	jobID := awardedJob(t, pool, customer, provider)

	svc, _, objects := driverProofService()
	deliverable(t, pool, svc, provider, jobID)

	key := objects.holding("proof/"+jobID.String()+"/019bd7a1-2c44-7f10-9a2c-3d4e5f607182", aPhotograph())
	proof, err := svc.VerifyProof(t.Context(), pool, provider, jobID, key)
	if err != nil {
		t.Fatalf("verifying the photograph: %v", err)
	}

	record, _, err := recordMilestone(t, pool, svc, provider, jobID, Recording{
		Milestone:     MilestoneDelivered,
		Key:           "comp1-delivered",
		Proof:         proof,
		RecipientName: "R. Chen",
		DeliveryNote:  "Left with reception, signed for",
	})
	if err != nil {
		t.Fatalf("recording the delivery: %v", err)
	}

	var recipient, note string
	if err := pool.QueryRow(t.Context(),
		`SELECT recipient_name, delivery_note FROM milestones WHERE id = $1`, record.ID).
		Scan(&recipient, &note); err != nil {
		t.Fatalf("reading the delivered row: %v", err)
	}
	if recipient != "R. Chen" || note != "Left with reception, signed for" {
		t.Fatalf("the row holds %q / %q", recipient, note)
	}

	// And they come back out through the domain, so a reader that never touches SQL sees them.
	if record.RecipientName != "R. Chen" || record.DeliveryNote != "Left with reception, signed for" {
		t.Fatalf("the record holds %q / %q", record.RecipientName, record.DeliveryNote)
	}
}

// TestADeliveryWithoutTheFieldSetIsRefused is `Docs/01` §4.4's requirement, one field at a time.
//
// Each case names the field, because that is the layer's whole job: `ck_milestones_delivery_details`
// would refuse the same rows with a constraint name and a 500, which `Docs/10` §4.6 says explains
// nothing.
//
// **A name that is only spaces is refused as required rather than stored**, which is
// `Recording.normalise` collapsing before `Recording.problems` judges — and is the case a length
// check alone would let through into a column the database then accepts.
func TestADeliveryWithoutTheFieldSetIsRefused(t *testing.T) {
	pool := pgtest.DB(t)
	customer := newAccount(t, pool, "comp2-c@example.com", "+61400000772", "customer")
	provider := newAccount(t, pool, "comp2-p@example.com", "+61400000773", "provider")
	jobID := awardedJob(t, pool, customer, provider)

	svc, _, _ := driverProofService()
	deliverable(t, pool, svc, provider, jobID)

	for _, tc := range []struct {
		name  string
		rec   Recording
		field string
	}{
		{"no recipient", Recording{RecipientName: "", DeliveryNote: "Left at the door"}, "recipient_name"},
		{"no note", Recording{RecipientName: "R. Chen", DeliveryNote: ""}, "delivery_note"},
		{"neither", Recording{}, "recipient_name"},
		{"a recipient of only spaces", Recording{RecipientName: "   ", DeliveryNote: "Left at the door"}, "recipient_name"},
		{"a note of only spaces", Recording{RecipientName: "R. Chen", DeliveryNote: "  "}, "delivery_note"},
		{"a recipient over the limit", Recording{RecipientName: strings.Repeat("x", maxRecipientName+1), DeliveryNote: "n"}, "recipient_name"},
		{"a note over the limit", Recording{RecipientName: "R. Chen", DeliveryNote: strings.Repeat("x", maxDeliveryNote+1)}, "delivery_note"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			rec := tc.rec
			rec.Milestone = MilestoneDelivered
			rec.Key = "comp2-" + tc.name
			rec.Exception = ExceptionRecipientObjected

			_, _, err := recordMilestone(t, pool, svc, provider, jobID, rec)

			var apiErr *httpx.Error
			if !errors.As(err, &apiErr) || apiErr.Code != httpx.CodeValidationFailed {
				t.Fatalf("got %v, want a validation failure naming %s", err, tc.field)
			}
			if field := fieldOf(t, apiErr); field != tc.field {
				t.Fatalf("the refusal names %q, want %q", field, tc.field)
			}
		})
	}

	if n := milestoneCount(t, pool, jobID); n != 3 {
		t.Fatalf("%d milestones on the job, want the 3 that drove it to In transit — a refused "+
			"delivery must write nothing", n)
	}
}

// TestOnlyADeliveredMilestoneCarriesThem is the other half of the rule, and the half a first draft
// leaves out.
//
// A recipient name on a `picked_up` is a handover that did not happen, and the column would happily
// hold it. Both the domain and `ck_milestones_delivery_details` refuse it.
func TestOnlyADeliveredMilestoneCarriesThem(t *testing.T) {
	pool := pgtest.DB(t)
	customer := newAccount(t, pool, "comp3-c@example.com", "+61400000774", "customer")
	provider := newAccount(t, pool, "comp3-p@example.com", "+61400000775", "provider")
	jobID := awardedJob(t, pool, customer, provider)

	svc, _, _ := driverProofService()

	for _, tc := range []struct {
		name  string
		rec   Recording
		field string
	}{
		{"a recipient on a pickup", Recording{RecipientName: "R. Chen"}, "recipient_name"},
		{"a note on a pickup", Recording{DeliveryNote: "Left with reception"}, "delivery_note"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			rec := tc.rec
			rec.Milestone = MilestoneEnRouteToPickup
			rec.Key = "comp3-" + tc.name

			_, _, err := recordMilestone(t, pool, svc, provider, jobID, rec)

			var apiErr *httpx.Error
			if !errors.As(err, &apiErr) || apiErr.Code != httpx.CodeValidationFailed {
				t.Fatalf("got %v, want a validation failure naming %s", err, tc.field)
			}
			if field := fieldOf(t, apiErr); field != tc.field {
				t.Fatalf("the refusal names %q, want %q", field, tc.field)
			}
		})
	}
}

// TestTheDeliveryFieldSetIsAlsoTheDatabases is the layer that does not depend on which function did
// the writing.
//
// The same argument SHIP-118 makes about `000605` and `000600` makes about its own unique index: the
// service checks so that a caller is told something useful, and the database enforces so that being
// told is not the mechanism. This writes raw SQL, which is the only way to reach the second layer.
func TestTheDeliveryFieldSetIsAlsoTheDatabases(t *testing.T) {
	pool := pgtest.DB(t)
	customer := newAccount(t, pool, "comp4-c@example.com", "+61400000776", "customer")
	provider := newAccount(t, pool, "comp4-p@example.com", "+61400000777", "provider")
	jobID := awardedJob(t, pool, customer, provider)

	for _, tc := range []struct {
		name      string
		milestone string
		recipient any
		note      any
	}{
		{"delivered with no recipient", "Delivered", nil, "Left with reception"},
		{"delivered with no note", "Delivered", "R. Chen", nil},
		{"delivered with neither", "Delivered", nil, nil},
		{"delivered with an empty recipient", "Delivered", "  ", "Left with reception"},
		{"a pickup carrying a recipient", "Picked up", "R. Chen", nil},
		{"a pickup carrying a note", "Picked up", nil, "Left with reception"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			id, _ := uuid.NewV7()
			_, err := pool.Exec(t.Context(), `
				INSERT INTO milestones
					(id, job_id, milestone, actor_type, actor_id, actor_recorded_at,
					 recipient_name, delivery_note)
				VALUES ($1, $2, $3, 'provider', $4, now(), $5, $6)`,
				id, jobID, tc.milestone, provider, tc.recipient, tc.note)

			if err == nil {
				t.Fatal("the database accepted a row Docs/01 §4.4 forbids, so the rule lives only " +
					"in the function that happened to be called")
			}
			if !strings.Contains(err.Error(), "ck_milestones_delivery_details") {
				t.Fatalf("refused by something other than the pairing constraint: %v", err)
			}
		})
	}
}

// TestTheDeliveryFieldLimitsAreOneLimit is Docs/10 §3.4's pairing for a bound rather than an
// enumeration.
//
// Two numbers stated in two languages is two numbers that drift, and the drift is silent in the
// worse direction: a Go bound looser than the column's turns a client's perfectly good input into a
// constraint name and a 500. The constraint definition is read out of the catalogue rather than out
// of the migration file, so a later `ALTER` that changed one and not the other is caught.
func TestTheDeliveryFieldLimitsAreOneLimit(t *testing.T) {
	pool := pgtest.DB(t)

	var definition string
	if err := pool.QueryRow(t.Context(), `
		SELECT pg_get_constraintdef(oid) FROM pg_constraint
		 WHERE conname = 'ck_milestones_delivery_details'`).Scan(&definition); err != nil {
		t.Fatalf("reading ck_milestones_delivery_details out of pg_constraint: %v", err)
	}

	// The bound is matched against the *column* rather than merely against the constraint text,
	// so that swapping the two numbers is caught. PostgreSQL normalises `BETWEEN 1 AND 120` into a
	// pair of comparisons when it renders the definition, which is why the expected string is
	// built from the catalogue's own spelling and not from the migration's.
	for column, want := range map[string]int{
		"recipient_name": maxRecipientName,
		"delivery_note":  maxDeliveryNote,
	} {
		bound := fmt.Sprintf("length(btrim(%s)) <= %d", column, want)
		if !strings.Contains(definition, bound) {
			t.Errorf("the constraint does not bound %s at the Go limit of %d (%q): %s",
				column, want, bound, definition)
		}
		if !strings.Contains(definition, fmt.Sprintf("length(btrim(%s)) >= 1", column)) {
			t.Errorf("the constraint does not refuse an empty %s: %s", column, definition)
		}
	}
}

// TestTheDriverIsToldWhenTheDeliveryIsFinished is the second clause of the *Done when*: "portal
// becomes read-only after".
//
// # Why the platform answers it rather than the page remembering
//
// A driver reloads. Component state does not survive that, `sessionStorage` would be the page
// inventing a fact about the delivery, and there is no driver-readable milestone list — `GET
// /v1/jobs/{id}/delivery/milestones` is `RequireUser`. So `delivered_at` is served, which is also
// where `Docs/07` §3 puts every question of this kind.
//
// Asserted over HTTP because the field's absence is `omitempty` and a struct comparison would not
// distinguish "absent" from "empty string" the way a client parsing JSON does.
func TestTheDriverIsToldWhenTheDeliveryIsFinished(t *testing.T) {
	pool := pgtest.DB(t)
	customer := newAccount(t, pool, "comp5-c@example.com", "+61400000778", "customer")
	provider := newAccount(t, pool, "comp5-p@example.com", "+61400000779", "provider")
	jobID := awardedJob(t, pool, customer, provider)
	_, grant, token := driverOnJob(t, pool, provider, jobID)

	svc, _, objects := driverProofService()
	router := newDriverRouterFor(t, pool, svc)

	// Before: no key at all, which is what keeps the closed-set guard in driverauth_test.go true
	// of an undelivered job.
	if _, ok := driverJobKeys(t, router, jobID, token.Value)["delivered_at"]; ok {
		t.Fatal("an undelivered job reports a delivered_at")
	}

	deliverable(t, pool, svc, provider, jobID)

	key := objects.holding("proof/"+jobID.String()+"/019bd7a1-2c44-7f10-9a2c-3d4e5f607182", aPhotograph())
	proof, err := svc.VerifyDriverProof(t.Context(), pool, grant, key)
	if err != nil {
		t.Fatalf("verifying the driver's photograph: %v", err)
	}
	if _, _, err := recordAsDriver(t, pool, svc, grant, Recording{
		Milestone:     MilestoneDelivered,
		Key:           "comp5-delivered",
		Proof:         proof,
		RecipientName: "R. Chen",
		DeliveryNote:  "Left with reception, signed for",
	}); err != nil {
		t.Fatalf("recording the delivery: %v", err)
	}

	// After: present, in UTC, and the link still opens the page — a driver may reopen it to check
	// what they recorded, and the field is presentation rather than authorisation.
	after := driverJobKeys(t, router, jobID, token.Value)
	delivered, ok := after["delivered_at"].(string)
	if !ok || delivered == "" {
		t.Fatalf("a delivered job reports %v for delivered_at", after["delivered_at"])
	}
	if !strings.HasSuffix(delivered, "Z") {
		t.Errorf("delivered_at = %q, want UTC", delivered)
	}
}

// newDriverRouterFor is [newDriverRouter] over a service the test built, so that the stubbed store
// used to verify a photograph is the same one the route reads through.
func newDriverRouterFor(t *testing.T, pool *pgxpool.Pool, svc *Service) http.Handler {
	t.Helper()

	handler, err := NewHandler(svc, pool, slog.New(slog.NewTextHandler(io.Discard, nil)))
	if err != nil {
		t.Fatalf("building the handler: %v", err)
	}

	mux := http.NewServeMux()
	mux.Handle("GET /v1/driver/jobs/{id}",
		RequireDriverToken(testDriverVerifier(t, clock.NewFixed(testDriverIssuedAt)))(handler.DriverJob()))
	return mux
}

// driverJobKeys is what `GET /v1/driver/jobs/{id}` answered, as a map, so a test can ask whether a
// key is present rather than whether a field is empty.
func driverJobKeys(t *testing.T, h http.Handler, jobID uuid.UUID, token string) map[string]any {
	t.Helper()

	rec := openLink(h, jobID, bearer(token))
	if rec.Code != http.StatusOK {
		t.Fatalf("the driver's own delivery answered %d: %s", rec.Code, rec.Body)
	}

	var body map[string]any
	if err := json.Unmarshal(rec.Body.Bytes(), &body); err != nil {
		t.Fatalf("the response is not JSON: %v", err)
	}
	return body
}

// fieldOf reads the first field name out of a validation failure.
func fieldOf(t *testing.T, err *httpx.Error) string {
	t.Helper()

	if len(err.Details) == 0 {
		t.Fatalf("the refusal carries no field errors: %v", err)
	}
	return err.Details[0].Field
}

File: apps/driver-portal/lib/completion.test.ts
import assert from "node:assert/strict";
import { register } from "node:module";
import { test } from "node:test";

import { openDelivery, type Delivery } from "./delivery.ts";
import { capturePhotograph, recordStep } from "./open.ts";

/**
 * **The delivery completion form, and the field `Docs/01` §4.4 required for eight waves before
 * anything could hold it** (SHIP-123).
 *
 * That paragraph requires four things of a delivered job — recipient name, delivery timestamp,
 * delivery note, and photo proof. Three were reachable before this ticket; `000605`'s own comment
 * named SHIP-123 as the one that would close the fourth and the fifth, and `Docs/11` §4 carried the
 * gap as documentation ahead of code.
 *
 * What is asserted here is the client's half: that both fields reach the platform on a `delivered`
 * and on nothing else, through both paths to it — the photograph and the reasoned exception — and
 * that `delivered_at` is read back so the page can stop offering the controls.
 *
 * **The last group is SHIP-131a's**, and it is here because this file already runs both paths to a
 * finished delivery with the platform's request bodies captured. `MilestoneRecording.reason` is a
 * third field, unrelated to the two above — the optional note **any** milestone may carry, which the
 * platform has accepted since SHIP-111 and no client has ever sent, while the customer's tracking
 * view rendered it. Every assertion is on the body that reached the platform, never on the form,
 * because a note the driver can see and the platform never receives is the defect the ticket exists
 * to close and a form-level assertion would pass straight through it.
 */

register("./alias-hooks.mts", import.meta.url);

const { POST: PRESIGN } = await import("../app/api/driver/jobs/[jobId]/proof-uploads/route.ts");
const { POST: RECORD } = await import("../app/api/driver/jobs/[jobId]/milestones/route.ts");
const { GET: READ } = await import("../app/api/driver/jobs/[jobId]/route.ts");

const ORIGIN = "https://portal.example";
const STORE = "https://objects.example/shipper-dev";
const JOB = "6f2a1f3c-9d4e-4b1a-8c7d-2e5f0a9b1c3d";
const OBJECT_KEY = `proof/${JOB}/019bd7a1-2c44-7f10-9a2c-3d4e5f607182`;

function segment(value: unknown): string {
  return Buffer.from(JSON.stringify(value)).toString("base64url");
}

const TOKEN = [
  segment({ alg: "EdDSA", typ: "JWT", kid: "driver-1" }),
  segment({ job_id: JOB, assignment_id: "aaaaaaaa-bbbb-4ccc-8ddd-eeeeeeeeeeee", aud: "driver" }),
  "c2lnbmF0dXJlLW5vdGhpbmctaGVyZS12ZXJpZmllcw",
].join(".");

const FINISHED = { recipientName: "R. Chen", deliveryNote: "Left with reception, signed for" };

/** The bodies this application sent to the platform, in order, parsed. */
type Bodies = Record<string, unknown>[];

/**
 * Run one page action in a browser that is not there, and keep the bodies that reached the platform.
 *
 * Both hops are real: the portal's own route handlers are the ones Next would run, and only the
 * platform and the store are stand-ins.
 */
async function inTheBrowser<T>(
  act: () => Promise<T>,
  delivered?: string,
): Promise<{ outcome: T; bodies: Bodies }> {
  const bodies: Bodies = [];
  const held = new Map<string, string>();

  const browser = {
    location: { pathname: `/j/${JOB}`, hash: `#${TOKEN}` },
    history: { replaceState: () => undefined },
    sessionStorage: {
      getItem: (key: string) => held.get(key) ?? null,
      setItem: (key: string, value: string) => void held.set(key, value),
      removeItem: (key: string) => void held.delete(key),
    },
  };

  const platform = (path: string, body: unknown): Response => {
    if (path.endsWith("/proof-uploads")) {
      const asked = body as { content_type: string; content_length: number };
      return Response.json({
        object_key: OBJECT_KEY,
        upload_url: `${STORE}/${OBJECT_KEY}?X-Amz-Signature=abc`,
        method: "PUT",
        content_type: asked.content_type,
        content_length: asked.content_length,
        expires_at: "2026-08-14T02:07:00.000Z",
      });
    }
    if (path.endsWith("/milestones")) {
      const asked = body as { milestone: string };
      return Response.json(
        {
          id: "01999999-0000-7000-8000-000000000001",
          job_id: JOB,
          milestone: asked.milestone,
          recorded_by: "driver",
          recorded_at: "2026-08-14T02:00:00.000Z",
          accepted_at: "2026-08-14T02:00:01.000Z",
        },
        { status: 201 },
      );
    }

    const view: Delivery = {
      job_id: JOB,
      assignment_id: "aaaaaaaa-bbbb-4ccc-8ddd-eeeeeeeeeeee",
      driver_name: "Jo Driver",
      assigned_at: "2026-08-13T04:00:00Z",
      link_expires_at: "2026-08-20T04:00:00Z",
    };
    // `delivered_at` is omitted entirely rather than sent empty, which is what `omitempty` does on
    // the platform side and is the case `isDelivery` has to keep accepting.
    if (delivered !== undefined) view.delivered_at = delivered;
    return Response.json(view);
  };

  const sending = async (input: string | URL | Request, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input instanceof URL ? input.href : input.url;

    if (url.startsWith(STORE)) return new Response(null, { status: 200 });

    if (!/^https?:/i.test(url)) {
      const address = new URL(url, ORIGIN);
      const params = {
        params: Promise.resolve({ jobId: decodeURIComponent(address.pathname.split("/")[4] ?? "") }),
      };
      if (address.pathname.endsWith("/proof-uploads")) {
        return PRESIGN(
          new Request(address, { method: "POST", headers: init?.headers, body: init?.body as BodyInit }),
          params,
        );
      }
      if (address.pathname.endsWith("/milestones")) {
        return RECORD(
          new Request(address, { method: "POST", headers: init?.headers, body: init?.body as BodyInit }),
          params,
        );
      }
      return READ(new Request(address, { headers: init?.headers }), params);
    }

    const body: unknown = init?.body === undefined ? null : JSON.parse(String(init.body));
    if (body !== null) bodies.push(body as Record<string, unknown>);
    return platform(new URL(url).pathname, body);
  };

  const globals = globalThis as unknown as { window?: unknown; fetch?: typeof fetch };
  const realFetch = globals.fetch;
  globals.window = browser;
  globals.fetch = sending as typeof fetch;

  try {
    return { outcome: await act(), bodies };
  } finally {
    globals.fetch = realFetch;
    delete globals.window;
  }
}

/** The recipient and the note reach the platform on the exception path. */
test("a delivery recorded with a reasoned exception carries the recipient and the note", async () => {
  const { outcome, bodies } = await inTheBrowser(() =>
    recordStep(JOB, "delivered", {
      evidence: { exception_reason: "camera_unavailable" },
      completion: FINISHED,
    }));

  assert.equal(outcome.kind, "recorded");
  assert.equal(bodies.length, 1);
  assert.deepEqual(bodies[0].proof, { exception_reason: "camera_unavailable" });
  assert.equal(bodies[0].recipient_name, "R. Chen");
  assert.equal(bodies[0].delivery_note, "Left with reception, signed for");
});

/** And on the photographed path, which is three requests rather than one. */
test("a photographed delivery carries them on the recording, not on the presign", async () => {
  const photograph = new Blob([new Uint8Array(1874)], { type: "image/jpeg" });
  const { outcome, bodies } = await inTheBrowser(() =>
    capturePhotograph(JOB, "delivered", photograph, { completion: FINISHED }));

  assert.equal(outcome.kind, "recorded");
  assert.equal(bodies.length, 2, "the presign and the recording");

  // The presign describes bytes and nothing else. A recipient name in it would be personal data on
  // a request whose whole subject is a media type and a length.
  assert.deepEqual(Object.keys(bodies[0]).sort(), ["content_length", "content_type"]);

  assert.equal(bodies[1].milestone, "delivered");
  assert.equal(bodies[1].recipient_name, "R. Chen");
  assert.equal(bodies[1].delivery_note, "Left with reception, signed for");
  assert.deepEqual(bodies[1].proof, { object_key: OBJECT_KEY });
});

/**
 * Every other milestone sends neither, because the platform refuses both on any milestone but
 * `delivered` — a recipient name on a pickup is a handover that did not happen.
 */
test("an ordinary milestone carries neither field", async () => {
  const { bodies } = await inTheBrowser(() => recordStep(JOB, "picked_up"));

  assert.equal(bodies.length, 1);
  assert.equal("recipient_name" in bodies[0], false);
  assert.equal("delivery_note" in bodies[0], false);
  assert.deepEqual(Object.keys(bodies[0]).sort(), ["milestone", "recorded_at"]);
});

/**
 * **SHIP-131a's central assertion, and the one the mutation is aimed at.** A note typed on the
 * portal has to arrive as `MilestoneRecording.reason`. Asserted on the body the platform received,
 * so a build that renders the field and drops it from the request fails here.
 */
test("a note typed on the portal arrives as `reason`", async () => {
  const { outcome, bodies } = await inTheBrowser(() =>
    recordStep(JOB, "picked_up", { note: "Nobody at the gate, returning at four" }));

  assert.equal(outcome.kind, "recorded");
  assert.equal(bodies.length, 1);
  assert.equal(bodies[0].reason, "Nobody at the gate, returning at four");
  assert.deepEqual(Object.keys(bodies[0]).sort(), ["milestone", "reason", "recorded_at"]);
});

/**
 * On the exception path it sits **beside** the selected reason rather than instead of one.
 *
 * `Docs/01` §4.4's list is closed so that `Docs/04` §5's delivery-exception queue can group it, and
 * `ck_proofs_exception_reason` pairs with that. The note is what says which of the three it actually
 * was — "the recipient asked me not to photograph their door" is the sentence that turns a queue
 * entry into a decision. Both fields, one body.
 */
test("a note goes beside the exception reason and never in place of it", async () => {
  const { outcome, bodies } = await inTheBrowser(() =>
    recordStep(JOB, "delivered", {
      evidence: { exception_reason: "recipient_objected" },
      completion: FINISHED,
      note: "The recipient asked me not to photograph their door.",
    }));

  assert.equal(outcome.kind, "recorded");
  assert.equal(bodies.length, 1);
  assert.deepEqual(bodies[0].proof, { exception_reason: "recipient_objected" });
  assert.equal(bodies[0].reason, "The recipient asked me not to photograph their door.");
});

/** And on the photographed path — on the recording, never on the presign. */
test("a note reaches the recording and not the request for somewhere to put a photograph", async () => {
  const photograph = new Blob([new Uint8Array(1874)], { type: "image/jpeg" });
  const { outcome, bodies } = await inTheBrowser(() =>
    capturePhotograph(JOB, "delivered", photograph, {
      completion: FINISHED,
      note: "Left inside the roller door, out of the rain.",
    }));

  assert.equal(outcome.kind, "recorded");
  assert.equal(bodies.length, 2, "the presign and the recording");

  // The presign's whole subject is a media type and a length. A driver's sentence about the
  // delivery on it would be personal data on a request that has no business carrying any.
  assert.deepEqual(Object.keys(bodies[0]).sort(), ["content_length", "content_type"]);
  assert.equal(bodies[1].reason, "Left inside the roller door, out of the rain.");
});

/**
 * Optional in the strong sense: an empty note sends **no key**, not an empty string.
 *
 * Whitespace rather than nothing, because that is the case a form actually produces. The platform
 * collapses whitespace and then bounds what is left, so `"reason": "  "` is refused as too short —
 * and `"reason": ""` would be worse than a refusal, because the customer's tracking view branches on
 * the field being *present* and would draw a blank line under their latest update.
 */
test("a note of nothing but spaces is not sent at all", async () => {
  const { bodies } = await inTheBrowser(() => recordStep(JOB, "in_transit", { note: "   " }));

  assert.equal(bodies.length, 1);
  assert.equal("reason" in bodies[0], false);
  assert.deepEqual(Object.keys(bodies[0]).sort(), ["milestone", "recorded_at"]);
});

/**
 * A delivery that has not been delivered carries no `delivered_at`, and the page still reads it.
 *
 * This is the case the field's `omitempty` produces, and it is the reason `isDelivery` deliberately
 * does not require it: an absent `delivered_at` is an ordinary delivery in progress, not a broken
 * contract.
 */
test("a delivery in progress is read without a delivered_at", async () => {
  const { outcome } = await inTheBrowser(() => openDelivery(JOB, TOKEN));

  assert.equal(outcome.kind, "delivery");
  assert.equal(outcome.kind === "delivery" && outcome.delivery.delivered_at, undefined);
});

/**
 * A finished delivery reports when, which is the whole of how the page becomes read-only after a
 * reload — component state does not survive one, and the alternative would be this page inventing a
 * fact about the delivery rather than reading the platform's.
 */
test("a finished delivery is read back with the time it was delivered", async () => {
  const { outcome } = await inTheBrowser(
    () => openDelivery(JOB, TOKEN),
    "2026-08-14T06:22:11.000Z",
  );

  assert.equal(outcome.kind, "delivery");
  assert.equal(
    outcome.kind === "delivery" && outcome.delivery.delivered_at,
    "2026-08-14T06:22:11.000Z",
  );
});
